### PR TITLE
fix(forum): workaround fix for URI length error - help button

### LINKED ISSUE #18005: Ask for Help - Learn app to Forum Post URI length exceeds limit
#### Describe your problem and - if possible - how to reproduce it
Hi
I’ve found a bug in the given page. When I click on the “Ask for Help” Button and try to create help post in the forum, it gives a data transmission error as follows:

3c51d48a0b8cdd296aee2f1dd16ca3cd0ae27c6a png

I’ve tested this in Chrome and Mozilla Firefox. 

#### Add a Link to the page with the problem
Link to the challenge:
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/functional-programming/use-the-map-method-to-extract-data-from-an-array/

#### Tell us about your browser and operating system

* Browser Name: Google Chrome and Mozilla Firefox Quantum
* Browser Version: Chrome version- Version 68.0.3440.84 (Official Build) (64-bit)
                               Firefox version-  Version 61.0.1 (64-bit)
* Operating System: Windows 10 and Ubuntu 16.04 LTS

### PR DESCRIPTION
**Pull request by jragard**
The error is occurring because the user's current code is being sent as an encoded query parameter in a get request to the forum.  The standard default character limit for URI's in most browsers is 2083...I tested it out with a test URI and discovered that the limit set by the forum's backend is currently 8,138....so it looks like this has been an ongoing problem, and the character limit has been raised before as a hacky workaround, but this info should be sent via post request.  I fixed this issue by constructing a default URI and an alternative URI with the alternate data suggested by @RandellDawson in the following thread. 

If the default URI has a length of over 8,000 characters, the alternative URI is used to get to the forum.

I also added a line to eslintrc.json to allow commiting with 'unable to resolve path to module' linter warning...I was unable to commit because the linter couldn't resolve the import statement on line 1 (which is obviously needed).

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #18005

### CODE REVIEW

**Comment by raisedadead** [comment]
> I also added a line to eslintrc.json to allow commiting with 'unable to resolve path to module' linter warning...I was unable to commit because the linter couldn't resolve the import statement on line 1 (which is obviously needed).

You should not have needed to do this. The `import` should have worked, it's possible something is broken somewhere else.

Will investigate and let you know.

**Comment by jragard** [comment]
I also probably should have specified that the import flagged by the linter
was '  import { ofType } from 'redux-observable';  '....which I didn't
touch in my changes.  And I know it was being successfully imported in the
module because `ofType` was defined....I also noticed while playing around
in another unrelated module investigating a separate issue, when I tried to
commit a change in that file, the linter complained about `  import React
from 'react';  `, which I also didn't change at all.  And I know React was
successfully imported, because it still worked in the module...so, it seems
to me that the imports are happening successfully, but the linter still
doesn't like them for some reason...unless there is some other weird thing
going on that is over my head

**Comment by jragard** [comment]
One more thing.  I also realized that, you can add the "import/no-unresolved
": "off" line in eslintrc.json and git add any other changes (without
adding eslintrc.json), and it allows you to commit without throwing the
import warning.  Sorry, I should have probably raised that issue with you
before submitting a PR...but it seems like a really weird linter warning,
considering all the evidence that the imports are actually working fine.
Let me know what you discover

**Comment by ojeytonwilliams** [comment]
Hey @jragard I wasn't able to reproduce the linting issue.  Please could you remove the `"import/no-unresolved": "off"` line and then go to the repository and run 

```
npm run clean
rm -rf node_modules
npm ci
npm run lint
```

and then let me know if there are still any linting errors.

**Comment by jragard** [comment]
Hi @ojeytonwilliams, thanks for the help.  I'm still having some troubles...npm run clean isn't working correctly for me.  Below is the complete log from running the command:

```
0 info it worked if it ends with ok
1 verbose cli [ '/home/jragard/.nvm/versions/node/v10.16.0/bin/node',
1 verbose cli   '/home/jragard/.nvm/versions/node/v10.16.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'clean' ]
2 info using npm@6.9.0
3 info using node@v10.16.0
4 verbose run-script [ 'preclean', 'clean', 'postclean' ]
5 info lifecycle @freecodecamp/freecodecamp@0.0.1~preclean: @freecodecamp/freecodecamp@0.0.1
6 info lifecycle @freecodecamp/freecodecamp@0.0.1~clean: @freecodecamp/freecodecamp@0.0.1
7 verbose lifecycle @freecodecamp/freecodecamp@0.0.1~clean: unsafe-perm in lifecycle true
8 verbose lifecycle @freecodecamp/freecodecamp@0.0.1~clean: PATH: /home/jragard/.nvm/versions/node/v10.16.0/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/jragard/Desktop/openSource/freeCodeCamp/node_modules/.bin:/home/jragard/.nvm/versions/node/v10.16.0/bin:/home/jragard/.pyenv/plugins/pyenv-virtualenv/shims:/home/jragard/.pyenv/shims:/home/jragard/.pyenv/bin:/home/jragard/.pyenv/plugins/pyenv-virtualenv/shims:/home/jragard/.pyenv/shims:/home/jragard/.pyenv/bin:/home/jragard/.nix-profile/bin:/home/jragard/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/home/jragard/.local/bin:/home/jragard/.local/bin
9 verbose lifecycle @freecodecamp/freecodecamp@0.0.1~clean: CWD: /home/jragard/Desktop/openSource/freeCodeCamp
10 silly lifecycle @freecodecamp/freecodecamp@0.0.1~clean: Args: [ '-c', 'npm-run-all clean:gatsby-site clean:packages' ]
11 silly lifecycle @freecodecamp/freecodecamp@0.0.1~clean: Returned: code: 1  signal: null
12 info lifecycle @freecodecamp/freecodecamp@0.0.1~clean: Failed to exec clean script
13 verbose stack Error: @freecodecamp/freecodecamp@0.0.1 clean: `npm-run-all clean:gatsby-site clean:packages`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/home/jragard/.nvm/versions/node/v10.16.0/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:198:13)
13 verbose stack     at ChildProcess.<anonymous> (/home/jragard/.nvm/versions/node/v10.16.0/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:198:13)
13 verbose stack     at maybeClose (internal/child_process.js:982:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
14 verbose pkgid @freecodecamp/freecodecamp@0.0.1
15 verbose cwd /home/jragard/Desktop/openSource/freeCodeCamp
16 verbose Linux 4.15.0-52-generic
17 verbose argv "/home/jragard/.nvm/versions/node/v10.16.0/bin/node" "/home/jragard/.nvm/versions/node/v10.16.0/bin/npm" "run" "clean"
18 verbose node v10.16.0
19 verbose npm  v6.9.0
20 error code ELIFECYCLE
21 error errno 1
22 error @freecodecamp/freecodecamp@0.0.1 clean: `npm-run-all clean:gatsby-site clean:packages`
22 error Exit status 1
23 error Failed at the @freecodecamp/freecodecamp@0.0.1 clean script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

After running this command, I still tried to go forward with the rest a few times...I have no trouble removing node_modules, or with `npm ci`.  Just to be thorough, I will paste the full output from `npm ci` below as well.  I don't know if it's worth mentioning that I have been working with the docker build?

`npm ci` output:

```
jragard@jragard-ThinkPad-T440:~/Desktop/openSource/freeCodeCamp$ npm ci

> fsevents@1.2.9 install /home/jragard/Desktop/openSource/freeCodeCamp/node_modules/fsevents
> node install


> husky@1.3.1 install /home/jragard/Desktop/openSource/freeCodeCamp/node_modules/husky
> node husky install

husky > setting up git hooks
husky > done

> @freecodecamp/freecodecamp@0.0.1 postinstall /home/jragard/Desktop/openSource/freeCodeCamp
> npm run bootstrap


> @freecodecamp/freecodecamp@0.0.1 bootstrap /home/jragard/Desktop/openSource/freeCodeCamp
> lerna bootstrap --ci

lerna notice cli v3.14.1
lerna info versioning independent
lerna info ci enabled
lerna info Bootstrapping 11 packages
lerna info Symlinking packages and binaries
lerna success Bootstrapped 11 packages
added 1158 packages in 15.033s
```

And then finally, after running the previous commands, the terminal output from `npm run lint`, followed by the full log generated by npm.

Terminal Output:

```
jragard@jragard-ThinkPad-T440:~/Desktop/openSource/freeCodeCamp$ npm run lint

> @freecodecamp/freecodecamp@0.0.1 lint /home/jragard/Desktop/openSource/freeCodeCamp
> eslint .


/home/jragard/Desktop/openSource/freeCodeCamp/api-server/common/models/user.js
  10:20  error  Unable to resolve path to module 'moment'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/api-server/server/boot/challenge.js
  11:26  error  Unable to resolve path to module 'mongodb'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/api-server/server/middlewares/error-reporter.js
  2:21  error  Unable to resolve path to module 'rollbar'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/api-server/server/utils/about.js
  5:18  error  Unable to resolve path to module 'googleapis'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/api-server/server/utils/rx.js
  2:20  error  Unable to resolve path to module 'moment'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/client/src/components/profile/components/Camper.js
  5:25  error  Unable to resolve path to module '@fortawesome/free-solid-svg-icons'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/client/src/components/profile/components/SocialIcons.js
  10:24  error  Unable to resolve path to module '@fortawesome/free-solid-svg-icons'  import/no-unresolved

/home/jragard/Desktop/openSource/freeCodeCamp/client/src/pages/index.js
  12:28  error  Unable to resolve path to module '@fortawesome/free-solid-svg-icons'  import/no-unresolved

✖ 8 problems (8 errors, 0 warnings)

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @freecodecamp/freecodecamp@0.0.1 lint: `eslint .`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @freecodecamp/freecodecamp@0.0.1 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jragard/.npm/_logs/2019-06-30T01_04_55_814Z-debug.log
```

Full log:

```
0 info it worked if it ends with ok
1 verbose cli [ '/home/jragard/.nvm/versions/node/v10.16.0/bin/node',
1 verbose cli   '/home/jragard/.nvm/versions/node/v10.16.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'lint' ]
2 info using npm@6.9.0
3 info using node@v10.16.0
4 verbose run-script [ 'prelint', 'lint', 'postlint' ]
5 info lifecycle @freecodecamp/freecodecamp@0.0.1~prelint: @freecodecamp/freecodecamp@0.0.1
6 info lifecycle @freecodecamp/freecodecamp@0.0.1~lint: @freecodecamp/freecodecamp@0.0.1
7 verbose lifecycle @freecodecamp/freecodecamp@0.0.1~lint: unsafe-perm in lifecycle true
8 verbose lifecycle @freecodecamp/freecodecamp@0.0.1~lint: PATH: /home/jragard/.nvm/versions/node/v10.16.0/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/jragard/Desktop/openSource/freeCodeCamp/node_modules/.bin:/home/jragard/.nvm/versions/node/v10.16.0/bin:/home/jragard/.pyenv/plugins/pyenv-virtualenv/shims:/home/jragard/.pyenv/shims:/home/jragard/.pyenv/bin:/home/jragard/.pyenv/plugins/pyenv-virtualenv/shims:/home/jragard/.pyenv/shims:/home/jragard/.pyenv/bin:/home/jragard/.nix-profile/bin:/home/jragard/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/home/jragard/.local/bin:/home/jragard/.local/bin
9 verbose lifecycle @freecodecamp/freecodecamp@0.0.1~lint: CWD: /home/jragard/Desktop/openSource/freeCodeCamp
10 silly lifecycle @freecodecamp/freecodecamp@0.0.1~lint: Args: [ '-c', 'eslint .' ]
11 silly lifecycle @freecodecamp/freecodecamp@0.0.1~lint: Returned: code: 1  signal: null
12 info lifecycle @freecodecamp/freecodecamp@0.0.1~lint: Failed to exec lint script
13 verbose stack Error: @freecodecamp/freecodecamp@0.0.1 lint: `eslint .`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/home/jragard/.nvm/versions/node/v10.16.0/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:198:13)
13 verbose stack     at ChildProcess.<anonymous> (/home/jragard/.nvm/versions/node/v10.16.0/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:198:13)
13 verbose stack     at maybeClose (internal/child_process.js:982:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
14 verbose pkgid @freecodecamp/freecodecamp@0.0.1
15 verbose cwd /home/jragard/Desktop/openSource/freeCodeCamp
16 verbose Linux 4.15.0-52-generic
17 verbose argv "/home/jragard/.nvm/versions/node/v10.16.0/bin/node" "/home/jragard/.nvm/versions/node/v10.16.0/bin/npm" "run" "lint"
18 verbose node v10.16.0
19 verbose npm  v6.9.0
20 error code ELIFECYCLE
21 error errno 1
22 error @freecodecamp/freecodecamp@0.0.1 lint: `eslint .`
22 error Exit status 1
23 error Failed at the @freecodecamp/freecodecamp@0.0.1 lint script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

**Comment by ojeytonwilliams** [comment]
@jragard Sorry, I should have asked if you're using Docker!  There are a couple of things we can try.

```
npm ci
npm run lint
```
which should output
```
> @freecodecamp/freecodecamp@0.0.1 lint /home/oliver/Code/fcc/freeCodeCamp
> eslint .
```
or, assuming everything is fine when you run `npm run docker:develop`
```
docker-compose run freecodecamp npm run lint
```
which should output
```
Starting freecodecamp_mailhog_1 ... done
Starting freecodecamp_db_1      ... done

> @freecodecamp/freecodecamp@0.0.1 lint /app
> eslint .
```

**Comment by jragard** [comment]
@ojeytonwilliams No worries!  `docker-compose run freecodecamp npm run lint` gave me the expected output.  I went ahead and deleted the import rule and committed last night, looks like it's pending review.  Thanks again for your help.

File: client/src/templates/Challenges/redux/create-question-epic.js [hunk 1]
```diff
@@ -44,28 +44,63 @@ function createQuestionEpic(action$, state$, { window }) {
...
+      ].join('');
+
+      const category = window.encodeURIComponent(
+        helpCategory[challengeType] || 'Help'
       );
+
+      const studentCode = window.encodeURIComponent(textMessage);
+      const altStudentCode = window.encodeURIComponent(altTextMessage);
+
+      const defaultURI = `https://forum.freecodecamp.org/new-topic?category=${category}
```
Comment (raisedadead) [inline client/src/templates/Challenges/redux/create-question-epic.js]:
```suggestion
      const defaultURI = `https://www.freecodecamp.org/forum/new-topic?category=${category}
```

File: client/src/templates/Challenges/redux/create-question-epic.js [hunk 2]
```diff
@@ -44,28 +44,63 @@ function createQuestionEpic(action$, state$, { window }) {
...
+                          &title=&body=${studentCode}`;
+      const altURI = `https://forum.freecodecamp.org/new-topic?category=${category}
```
Comment (raisedadead) [inline client/src/templates/Challenges/redux/create-question-epic.js]:
```suggestion
      const altURI = `https://www.freecodecamp.org/forum/new-topic?category=${category}
```

**Review by raisedadead — CHANGES_REQUESTED** [review]
The forum has moved, and while we are at this lets update the links. That said, I would really want to pick this up from the environment variables as we do for the nav component.

**Pull request closed without merging**